src/cpp/common/tls_credentials_options_util.cc

```diff
@@ -0,0 +1,144 @@
+      grpc_tls_key_materials_config_create();
+  ::grpc_core::InlinedVector<::grpc_core::PemKeyCertPair, 1>
+      c_pem_key_cert_pair_list;
+  for (auto key_cert_pair = config->pem_key_cert_pair_list().begin();
+       key_cert_pair != config->pem_key_cert_pair_list().end();
+       key_cert_pair++) {
+    grpc_ssl_pem_key_cert_pair* ssl_pair =
+        (grpc_ssl_pem_key_cert_pair*)gpr_malloc(
+            sizeof(grpc_ssl_pem_key_cert_pair));
+    ssl_pair->private_key = gpr_strdup(key_cert_pair->private_key.c_str());
+    ssl_pair->cert_chain = gpr_strdup(key_cert_pair->cert_chain.c_str());
+    ::grpc_core::PemKeyCertPair c_pem_key_cert_pair =
+        ::grpc_core::PemKeyCertPair(ssl_pair);
+    c_pem_key_cert_pair_list.push_back(::std::move(c_pem_key_cert_pair));
+  }
+  ::grpc_core::UniquePtr<char> c_pem_root_certs(
+      gpr_strdup(config->pem_root_certs().c_str()));
+  c_config->set_key_materials(std::move(c_pem_root_certs),
+                              std::move(c_pem_key_cert_pair_list));
+  c_config->set_version(config->version());
+  return c_config;
+}
+
+/** Converts the C key materials config to a Cpp key materials config; it
+ * allocates memory for the Cpp config. **/
+std::shared_ptr<TlsKeyMaterialsConfig> ConvertToCppKeyMaterialsConfig(
+    const grpc_tls_key_materials_config* config) {
+  if (config == nullptr) {
+    return nullptr;
+  }
+  std::shared_ptr<TlsKeyMaterialsConfig> cpp_config(
+      new TlsKeyMaterialsConfig());
+  std::vector<TlsKeyMaterialsConfig::PemKeyCertPair> cpp_pem_key_cert_pair_list;
+  grpc_tls_key_materials_config::PemKeyCertPairList pem_key_cert_pair_list =
+      config->pem_key_cert_pair_list();
+  for (size_t i = 0; i < pem_key_cert_pair_list.size(); i++) {
+    ::grpc_core::PemKeyCertPair key_cert_pair = pem_key_cert_pair_list[i];
+    TlsKeyMaterialsConfig::PemKeyCertPair p = {key_cert_pair.private_key(),
+                                               key_cert_pair.cert_chain()};
+    cpp_pem_key_cert_pair_list.push_back(::std::move(p));
+  }
+  cpp_config->set_key_materials(std::move(config->pem_root_certs()),
+                                std::move(cpp_pem_key_cert_pair_list));
+  cpp_config->set_version(config->version());
+  return cpp_config;
+}
+
+/** The C schedule and cancel functions for the credential reload config.
+ * They populate a C credential reload arg with the result of a C++ credential
+ * reload schedule/cancel API. **/
+int TlsCredentialReloadConfigCSchedule(void* config_user_data,
+                                       grpc_tls_credential_reload_arg* arg) {
+  if (arg == nullptr || arg->config == nullptr ||
+      arg->config->context() == nullptr) {
+    gpr_log(GPR_ERROR, "credential reload arg was not properly initialized");
+    return 1;
+  }
+  TlsCredentialReloadConfig* cpp_config =
+      static_cast<TlsCredentialReloadConfig*>(arg->config->context());
+  TlsCredentialReloadArg cpp_arg(arg);
+  return cpp_config->Schedule(&cpp_arg);
```

Comment:
I don't think this is a bug, assuming you are referring to the use of `&cpp_arg' in the line `cpp_config->Schedule(&cpp_arg)'. The reason is the following: 

-the function `cpp_config->Schedule` calls the `Schedule` function from the user-provided `CredentialReloadInterface` (or derived class, thereof); 

-this user-provided function takes an argument of type `TlsCredentialReloadArg` and performs some operations on this argument, but all that these operations do is modify the underlying C arg, of type `grpc_tls_credential_reload_arg`; 

-when the destructor of the `TlsCredentialReloadArg` is called, it does not free this underlying C arg;

-thus, the line `cpp_config->Schedule(&cpp_arg)` modifies the variable `arg` using the user-provided `Schedule` function, and using the local `cpp_arg` is just the way that `arg` is passed to the `Schedule` function.

Does that make sense? It's possible I've missed something, but this bug (at least as I understood it) doesn't show up in the unit tests. Also if I misunderstood your comment, please let me know!

Reply:
As I said I did not look at the test :) 
From the API point of view, how do I know which reload to cancel if the two pointers are not the same?
Let's say I am an application implementing the interface, I would likely to use the passed in arg to look up pending reloads and then cancel the proper one. If you just pass in a random pointer with shared underlying data, I lose the ability to do that.

I have been assuming the API says you schedule a reload with `arg` and then later it can be cancelled by calling `Cancel(arg)`. Hence I called it a bug.

Reply:
Ok I think I understand your point. Given my understanding, I've made the arguments of the Schedule and Cancel functions, and the `TlsCredentialReloadArg` above into `unique_ptr`'s, which I think takes care of the issue. Please let me know if you agree with this fix.

Reply:
I am not sure how it solves the problem. Basically what I am curious about is the scenario where two `Schedule` are called with different args and one of them needs to be cancelled. Is that a valid use case at all or does it violate the API contract? If it is valid, then in the `Cancel` call we need to somehow relate the Arg in the `Cancel` to one of the previous `Schedule` call.

Reply:
AFAIK yes, that is a valid use case. But to the user, all they have are the two instances, say `arg1` and `arg2`, of `TlsCredentialReloadArg`, and both belong to the same `TlsCredentialReloadConfig` object `config`. If we have called `Schedule` with both args and we would like to cancel `arg1`, then we can call `config.Cancel(arg1)` to cancel the first one. Why does this necessitate any further connection between what happens in the `Schedule` and `Cancel` functions?

As an aside, I just want to point out that the particular function, `TlsCredentialReloadCSchedule`, that this discussion is taking place under is completely internal - it is used only to create a `grpc_tls_credential_reload_config` instance given a user-provided `TlsCredentialReloadConfig` instance. It is only exposed for testing purposes (and I will add further comments to make this evident).

My apologies if I'm missing something straightforward here. The reason I added `unique_ptr`'s in the previous commit was to avoid any issues that might come up if schedule calls occurred in different threads, just to guard against anything funny occurring with local variables in that instance.

Reply:
I feel like there is something missing on either side. Could you point me to a doc describing how this API is intended to be used? Or we can have a meeting to discuss the issues. Thanks.

Reply:
@yihuazhang for reference.